### PR TITLE
Major event system improvement/refactoring. Now with Annotations

### PR DESCRIPTION
**Pull request by t3hk0d3**
Now using Annotations.
- Backward compatible.
- Prefomance friendly. PreEvent Switching. For each event handler own "personal" EventExecutor created, unlike #277, where rounting happens durning event (in onEvent). This mean event get routed before it has been actually called.
  But this affect only Annotated handlers (default one too). You still can use onEvent for universal event catching.
- Offers more flexiblility - onEvent or using Annotations

Annotations example:

```
public class MyListener implements Listener {
    @EventHandler(Type.BLOCK_DAMAGE)
    public void myBlockDamageEvent(BlockDamageEvent event) {
        // ... do your stuff ...
    }
}
```
- Correctly formatted :)

### CODE REVIEW

**Comment by tips48** [comment]
looks good!

**Comment by desmin88** [comment]
+1

**Comment by LadyCailin** [comment]
Why hasn't this been pulled yet?

**Comment by desmin88** [comment]
Probably won't ever be pulled.

**Comment by t3hk0d3** [comment]
At least their conscience would be dirty :)

**Comment by Majiir** [comment]
The sheer volume of code removed from `JavaPluginLoader` makes me happy.

**Comment by md-5** [comment]
Ummm Well to give this a snowballs chance in hell you MUST use @deprecated for the old ones, even though I sorta like it, I doubt it will get pulled :)

**Comment by Majiir** [comment]
@md-5: It doesn't look like this removes any old methods. It adds `@EventHandler` annotations instead of the massive `switch` for registering events. In other words, there's nothing to mark as `@Deprecated`.

**Comment by md-5** [comment]
I don't have time now, but looking quickly at the diff, I see a lot of red removals for Type.BLOCK_BREAK event registration. Anyway, I have only spent 3.2 seconds looking at it.

EDIT:A After another 4 seconds I realized he has just replaced the old switch with his new factory, so old plugins won't break?

File: src/main/java/org/bukkit/event/CustomEventListener.java [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
+
     public CustomEventListener() {}
+    
+    public void onEvent(Event event) {}
 
     /**
      * Called when a player joins a server
      *
      * @param event Relevant event details
      */
```
Comment (Majiir) [inline src/main/java/org/bukkit/event/CustomEventListener.java:16]: Why does the Javadoc describe `onCustomEvent` as a player join event?

**Comment by Majiir** [comment]
@md-5: Old plugins should work just fine.

In fact, if I understand this correctly, you _could_ add `@Deprecated` to all of the old event listeners to encourage plugin developers to use annotations instead. New Bukkit events don't even get an old listener because there's no plugin compatibility to worry about; they're just added to `Event.Type`.

**Comment by md-5** [comment]
Great.
It's a long shot to get pulled, @ deprecated might work if @dinnerbone or @evilseph likes this.
gtg

**Comment by Majiir** [comment]
Since `AnnotationHelper` contains only static methods, it should probably be declared `final` with a `private` (or `protected`) constructor.

**Comment by t3hk0d3** [comment]
@md-5 @Majiir Thank you guys, but this PR would be never pulled.

**Comment by md-5** [comment]
Thats what I said :)

**Comment by Dinnerbone** [comment]
I actually have this open in another tab to be looked at later.

**Comment by Dinnerbone** [comment]
We now have a new event system, thank you for the work but I'm afraid I must close this off.

**Comment by t3hk0d3** [comment]
@Dinnerbone Thank you. @zml2008's version are much better than this one.

**Pull request closed without merging**